### ISSUE TITLE
Why use .clamp(0).sqrt() on transformed input?

### ISSUE THREAD

**Issue by KristianMoellmann** [angus924/hydra#9]
Hi, 

I've been reading your paper and looking through the code, but I have difficulty understanding why you use `.clamp(0).sqrt()` on the transformed input data. See, e.g., line 94 in softmax.py:

`94    X_training_transform = transform.batch(X_training).clamp(0).sqrt()`

Have I missed something in the paper or somewhere else?

Thank you in advance for any help you can provide!

Best regards, Kristian

**Comment by angus924** [comment]
Hi @KristianMoellmann, thanks for your question.

Good question.

You'll notice that for smaller datasets, where we are using the ridge classifier, we use SparseScaler (in `hydra.py`) which does the same thing.

When training by gradient descent / minibatch, I currently implement this by "manually" calling clamp + sqrt on the output of the transform.  Actually, it would make more sense to simply add a class like BatchNorm1d to `model = nn.Sequential(nn.Linear(...))`, and this is in fact the approach I now take, although I haven't got around to updating the code in the repository yet.

In any case, let me explain what is going on.

The first thing to say is that if you are using the default parameters (k = 8, g = 64), clamp + sqrt is very unlikely to make any meaningful difference to the result.

The code in the SparseScaler class might be easier to read in this respect.

Basically, the problem we are trying to address is sparsity in the features.  Hydra works by, for each group of kernels, finding the top matching kernel at each timepoint.  For small groups (e.g., k = 8), most of the time you can probably expect every kernel to be the top match for at least some time points.

However, as the size of the group grows, there will be an increasing number of kernels which are the top match only rarely, or maybe even never (i.e., there is always at least one better match at each time point).  In other words, the resulting features (pooling over the kernel outputs) become increasingly sparse.

This in itself isn't an issue per se.  The problem is, we want to standardise the features before training the classifier (if using a model like the ridge classifier or logistic regression etc).  Very sparse features can become a problem.  E.g., you have a feature which is always zero in the training set but suddenly turns up for one or two instances in the test set.  Or, is zero for almost all time series in the training set.  Computing a meaningful mean + standard deviation (or any other statistic) becomes very tricky.  Standard deviation in particular can be problematic, as if it is underestimated, it can "explode" the feature values.

The approach we ended up taking here was to (a) only subtract the mean from non-zero values [basically, if a feature is sparse, leave it sparse]; (b) add an "epsilon" value to the computed standard deviation in inverse proportion to sparsity [i.e., if the feature is very sparse in the training set, force the denominator to be close to 1 when standardizing the features]; and (c) square root transform the features (performing clamp first just to avoid possible negative values in the features).  For (c), note that the features are basically counts, or something approximating counts, so they typically have quite a skewed distribution (essentially always positive, sometimes with large outlier values).

All of these together appear to largely avoid crazy outcomes when standardizing the features.

But, as I said, this is mostly to handle cases of large k (i.e., big groups of kernels).  It shouldn't make much difference if k is reasonably small.

I hope that helps a bit to answer your question.

Let me know if you have any other questions.

**Comment by KristianMoellmann** [comment]
Hi again

Thank you so much for your incredibly detailed and helpful response!

Your explanation makes perfect sense, and I thank you for taking the time to explain why such a transformation is beneficial for larger groups.

I don't have any more questions at the moment, but I might be back at a later stage :)

Your work is valued.

Best regards, Kristian
